12/umbraco-cms/reference/routing/custom-middleware.md

```diff
@@ -0,0 +1,70 @@
+- `PostPipeline`renders content. By using middleware, the response of content can be changed before being rendered.
+- `PreRouting`can be used to change the incoming URL/definers which URL should be served for the content
+- `PostRouting`can be used to change how the content should be displayed based on routing parameters.
+- `Endpoints` is the final step of the request pipeline where the content is rendered and returned on the requested browser.
+
+The addition of the `PreRouting` and `PostRouting` is to allow correctly configuring the Cross-Origin Resource Sharing (CORS) middleware. This is done using the `IUmbracoPipelineFilters` without having to use the `WithCustomMiddleware()`.
+
+- `IUmbracoPipelineFilter` is an interface in Umbraco that allows the creation of custom filters which then modifies the behavior of the request pipeline. It can be used to change different aspects of how Umbraco handles incoming requests, such as changing content or adding security checks.
+- `WithCustomMiddleware()` is a method that can be used in Umbraco for adding custom middleware. This includes some specific customizable instructions that run in the request processing pipeline. This method is often used in combination with IUmbracoPipelineFilter.
+
+{% hint style="warning" %}
+
+`WithCustomMiddleware` should only be used as a last resort, as Umbraco can break if you forget to add middleware or add them in the wrong order.  
+
+{% endhint %}
+
+## Configuring the Cross-Origin Resource Sharing (CORS) middleware
+
+Create a composer with the following:
+
+```csharp
+
+using Umbraco.Cms.Core.Composing;
+using Umbraco.Cms.Web.Common.ApplicationBuilder;
+
+public class CorsComposer : IComposer
+{
+    public const string AllowAnyOriginPolicyName = nameof(AllowAnyOriginPolicyName);
+
+    public void Compose(IUmbracoBuilder builder)
+        => builder.Services
+        .AddCors(options => options.AddPolicy(AllowAnyOriginPolicyName, policy => policy.AllowAnyOrigin()))
+        .Configure<UmbracoPipelineOptions>(options => options.AddFilter(new UmbracoPipelineFilter("Cors", postRouting: app => app.UseCors())))
+        // For testing only
+        .Configure<UmbracoPipelineOptions>(options => options.AddFilter(new UmbracoPipelineFilter("CorsTest", endpoints: app => app.UseEndpoints(endpoints =>
+        {
+            endpoints.MapGet("/echo", context => context.Response.WriteAsync("echo")).RequireCors(AllowAnyOriginPolicyName);
+            endpoints.MapGet("/echo2", context => context.Response.WriteAsync("echo2"));
+        }))));
+}
+```
+
+You should be able to request /echo from any origin, but get an error when trying to fetch /echo2. You can test this by pasting the following JavaScript code in your browser console when not on the local Umbraco website (such as Umbraco.com):
+
+```json
```

Comment:
Should this not be formatted as javascript and not JSON?

Reply:
makes sense, I have changed it to javascript 💪